mongo/single_result.go

```diff
@@ -22,11 +22,11 @@ var ErrNoDocuments = errors.New("mongo: no documents in result")
 // SingleResult represents a single document returned from an operation. If the operation resulted in an error, all
 // SingleResult methods will return that error. If the operation did not return any documents, all SingleResult methods
 // will return ErrNoDocuments.
-type SingleResult struct {
+type SingleResult[T bson.Raw | bson.RawArray] struct {
```

Comment:
T defines the return type of the `SingleResult.Raw` method.

Reply:
I'm concerned about the scope of the change required to add this type parameter. One of the challenges with generic types is that they require parameterizing all functions and types that use them, unless only a single type is ever used (see `verifySingleResult`).

An alternate approach that doesn't impact any existing types is to add a new type `DistinctResult` that holds a `SingleResult` and mostly reproduces the API of `SingleResult`, changing the return type of `Raw`. That creates fewer parameterized type changes in existing code and provides the ability to independently extend the APIs of `DistinctResult` and `SingleResult`.

E.g.
```go
type DistinctResult struct {
	sr *SingleResult
}

func (dr *DistinctResult) Decode(v interface{}) error { return dr.sr.Decode(v) }
func (dr *DistinctResult) Err() error                 { return dr.sr.Err() }

func (dr *DistinctResult) Raw() (bson.RawArray, error) {
	r, err := dr.sr.Raw()
	return bson.RawArray(r), err
}
```

Reply:
[Here](https://sourcegraph.com/search?q=context:global+lang:Go++-repo:%5Egithub%5C.com/mongodb/mongo-go-driver%24+repo:has.content%28go.mongodb.org/mongo-driver%29+-file:%28%5E%7C/%29vendor/+mongo.SingleResult&patternType=standard&case=yes&sm=0&groupBy=repo) are some examples of user code that would be impacted by adding a type parameter to `SingleResult`. They seem to fall in a few categories:
* Database client wrappers that return a `*mongo.SingleResult`.
* Interfaces used for mocking the `Collection` API.
* Adding helper functions for using a `*mongo.SingleResult`.

Reply:
I understand your point. It's worth noting that most of these examples will require updates due to GODRIVER-2696. However, I believe it's reasonable to minimize the impact where possible